Code/GraphMol/FileParsers/GeneralFileReader.h

```diff
@@ -0,0 +1,141 @@
+	std::string basename;  
+	//! Special case maegz.
+	//! NOTE: also supporting case-insensitive filesystems
+	if(boost::algorithm::iends_with(path, ".maegz")){
+		fileFormat = "mae";
+		compressionFormat = "gz";
+		return;
+	}
+	else if(boost::algorithm::iends_with(path, ".gz")){
+		compressionFormat = "gz";
+		basename = path.substr(0, path.size() - 3);
+	} else if( boost::algorithm::iends_with(path, ".zst") ||
+      boost::algorithm::iends_with(path, ".bz2") ||
+      boost::algorithm::iends_with(path, ".7z")){
+		throw std::invalid_argument("Unsupported compression extension");	
+	} else {
+ 		basename = path;
+		compressionFormat = "";
+	}
+	for (auto const& suffix: supportedFileFormats) {
+		 if (boost::algorithm::iends_with(basename, "." + suffix)) {
+				 fileFormat = suffix;
+				 return;
+		 }
+	}
+	throw std::invalid_argument("Unsupported structure or compression extension");
+}
+
+//! returns a MolSupplier object based on the file name instantiated
+//! with the relevant options provided in the SupplierOptions struct
+/*!
+    <b>Note:</b>
+      - the caller owns the memory and therefore the pointer must be deleted
+*/
+
+MolSupplier* getSupplier(const std::string& path,
+                         const struct SupplierOptions& opt) {
+  std::string fileFormat = "";
+  std::string compressionFormat = "";
+  //! get the file and compression format form the path
+  determineFormat(path, fileFormat, compressionFormat);
+  std::istream* strm;
+  if (compressionFormat.empty()) {
+    strm = new std::ifstream(path.c_str());
+  } else {
+    strm = new gzstream(path);
+  }
+
+  //! Dispatch to the appropriate supplier
+  if (fileFormat == "sdf") {
+		if(opt.numWriterThreads > 0){
+			MultithreadedSDMolSupplier* sdsup = new MultithreadedSDMolSupplier(
+				strm, true, opt.sanitize, opt.removeHs, opt.strictParsing, opt.numWriterThreads);
+			return sdsup;			
+		} 
+    ForwardSDMolSupplier* sdsup = new ForwardSDMolSupplier(
+        strm, true, opt.sanitize, opt.removeHs, opt.strictParsing);
+    return sdsup;
+  }
+
+  else if (fileFormat == "smi" || fileFormat == "csv" || fileFormat == "txt" ||
+					 fileFormat == "tsv") {
+		if(opt.numWriterThreads > 0){
+			MultithreadedSmilesMolSupplier* smsup = new MultithreadedSmilesMolSupplier(
+				strm, true, opt.delimiter, opt.smilesColumn,  opt.nameColumn, opt.titleLine,
+				opt.sanitize, opt.numWriterThreads);
+			return smsup;
+		}
+    SmilesMolSupplier* smsup =
+        new SmilesMolSupplier(strm, true, opt.delimiter, opt.smilesColumn,
+                              opt.nameColumn, opt.titleLine, opt.sanitize);
+    return smsup;
+  }
+
+  else if (fileFormat == "mae") {
+    MaeMolSupplier* maesup =
+        new MaeMolSupplier(strm, true, opt.sanitize, opt.removeHs);
+    return maesup;
+  }
+
+  else if (fileFormat == "tdt") {
+    TDTMolSupplier* tdtsup = new TDTMolSupplier(
+        strm, true, opt.nameRecord, opt.confId2D, opt.confId3D, opt.sanitize);
+    return tdtsup;
+  }
```

Comment:
It looks like we are using determineFormat to throw the exception here when we have a format we don't understand.  This is causing the compiler to think we have a control path with no return value (which is true).

Perhaps we should have determineFormat return true/false and deal with exceptions in the main body.

Also I would throw BadFileException as opposed to invalid_argument

Reply:
I modified the code to throw `BadFileException` instead of `std::invalid_argument`. However, I don't think it would be better to do error handling in the main body since we might not be able to get helpful error messages. The way `determineFormat` is written now, it terminates exactly when the file or compression format is invalid. Furthermore, we also know in which case it failed and therefore we are able to log out the precise reason (at least in the case when the compression formats are `.zst`, `.bz2` and `.7z`).